pulsar/impl_consumer.go

```diff
@@ -0,0 +1,256 @@
+			return nil, newError(ResultInvalidConfiguration, "Topic names list must be null when use topicsPattern")
+		}
+		// TODO: impl logic
+	} else if options.Topics != nil && len(options.Topics) > 1 {
+		// TODO: impl logic
+	} else if options.Topics != nil && len(options.Topics) == 1 || options.Topic != "" {
+		var singleTopicName string
+		if options.Topic != "" {
+			singleTopicName = options.Topic
+		} else {
+			singleTopicName = options.Topics[0]
+		}
+		return singleTopicSubscribe(client, options, singleTopicName)
+	}
+
+	return nil, newError(ResultInvalidTopicName, "topic name is required for consumer")
+}
+
+func singleTopicSubscribe(client *client, options *ConsumerOptions, topic string) (*consumer, error) {
+	c := &consumer{
+		topicName: topic,
+		queue:     make(chan ConsumerMessage, options.ReceiverQueueSize),
+	}
+
+	partitions, err := client.TopicPartitions(topic)
+	if err != nil {
+		return nil, err
+	}
+
+	numPartitions := len(partitions)
+	c.consumers = make([]Consumer, numPartitions)
+
+	type ConsumerError struct {
+		err       error
+		partition int
+		cons      Consumer
+	}
+
+	ch := make(chan ConsumerError, numPartitions)
+
+	for partitionIdx, partitionTopic := range partitions {
+		go func(partitionIdx int, partitionTopic string) {
+			cons, err := newPartitionConsumer(client, partitionTopic, options, partitionIdx)
+			ch <- ConsumerError{
+				err:       err,
+				partition: partitionIdx,
+				cons:      cons,
+			}
+		}(partitionIdx, partitionTopic)
+	}
+
+	for i := 0; i < numPartitions; i++ {
+		ce, ok := <-ch
+		if ok {
+			err = ce.err
+			c.consumers[ce.partition] = ce.cons
+		}
+	}
+
+	if err != nil {
+		// Since there were some failures, cleanup all the partitions that succeeded in creating the consumers
+		for _, consumer := range c.consumers {
+			if !util.IsNil(consumer) {
+				if err = consumer.Close(); err != nil {
+					panic("close consumer error, please check.")
+				}
+			}
+		}
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *consumer) Topic() string {
+	return c.topicName
+}
+
+func (c *consumer) Subscription() string {
+	return c.consumers[0].Subscription()
+}
+
+func (c *consumer) Unsubscribe() error {
+	var errMsg string
+	for _, c := range c.consumers {
+		if err := c.Unsubscribe(); err != nil {
+			errMsg += fmt.Sprintf("topic %s, subscription %s: %s", c.Topic(), c.Subscription(), err)
+		}
+	}
+	if errMsg != "" {
+		return errors.New(errMsg)
+	}
+	return nil
+}
+
+func (c *consumer) Receive(ctx context.Context) (Message, error) {
+	for _, pc := range c.consumers {
+		go func(pc Consumer) {
+			if err := pc.ReceiveAsync(ctx, c.queue); err != nil {
+				return
+			}
+		}(pc)
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case msg, ok := <-c.queue:
+		if ok {
+			return msg.Message, nil
+		}
+		return nil, errors.New("receive message error")
+	}
+}
+
+func (c *consumer) ReceiveAsync(ctx context.Context, msgs chan<- ConsumerMessage) error {
+	//TODO: impl logic
```

Comment:
please do the same thing for all the TODO items as what I suggested as above.

Reply:
The function will impl in https://github.com/apache/pulsar-client-go/issues/36